### ISSUE TITLE
Addressing SPARQL EXISTS errata

### ISSUE THREAD

**Issue by afs** [w3c/sparql-query#156]
#### Recap

[TPAC 2023 presentation](https://docs.google.com/presentation/d/16U0Z0HR6MwNWdBe-ctm3wLFd6wiKCum7DFRaTpxkkdM)

Issues: [sparql-query/issues for EXISTS](https://github.com/w3c/sparql-query/issues?q=is%3Aissue+is%3Aopen+EXISTS)

After TPAC 2023, [an email]](https://lists.w3.org/Archives/Public/public-rdf-star-wg/2023Sep/0036.html) was sent to interested parties.

#### Proposals

1:: Improved substitution
[SEP-0007/Substitution](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md)

2:: SemiJoin/Antijoin
https://w3c.github.io/sparql-exists/docs/sparql-exists.html#proposal-a

#### Proposal 1

Proposal 1 is based on errata for the ["Substitution" operation](https://www.w3.org/TR/sparql11-query/#defn_substitute).

Full details including relationship to errata: [SEP-0007/Substitution](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md).

#### Proposal 2

Proposal 2 is [SemiJoin](https://en.wikipedia.org/wiki/Relational_algebra#Semijoin_(%E2%8B%89_and_%E2%8B%8A))/[AntiJoin](https://en.wikipedia.org/wiki/Relational_algebra#Antijoin_(%E2%96%B7)).

SPARQL already has [`MINUS`](https://www.w3.org/TR/sparql11-query/#defn_algMinus) which is an antijoin with a special condition for the case of disjoint domains (a decision of the SPARQL 1.1 working group).

#### A way forward.

A compromise way forward:

1. Replace "Substitute" with the errata-derived fix [SEP-0007/Substitution](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md)

1. Plan for adding LATERAL, SEMIJOIN and ANTIJOIN (both pure forms) to the SPARQL language. This may have to be additional features added in the "new features" phase due to timing.

**Comment by pchampin** [comment]
This was discussed during the [rdf-star meeting on 26 September 2024](https://www.w3.org/2024/09/26-rdf-star-minutes.html#t01).

**Comment by pchampin** [comment]
This was discussed during the [rdf-star meeting on 26 September 2024](https://www.w3.org/2024/09/26-rdf-star-minutes.html#t05).

**Comment by pchampin** [comment]
This was discussed during the [rdf-star meeting on 26 September 2024](https://www.w3.org/2024/09/26-rdf-star-minutes.html#t01).

<details><summary><i>View the transcript</i></summary>

<h3>Un-star operation to support RDF Dataset Canonicalization? <a href="https://github.com/w3c/rdf-star-wg/issues/114" rel="noopener noreferrer">1</a></h3>
<p><cite>niklasl:</cite> I tried to summarize the two options on the table: unstar based on named graphs, or base on classical reification</p>
<p><cite>&lt;william_vw&gt;</cite> present_</p>
<p><cite>niklasl:</cite> also the question whether it is targetting canonicalization alone, or more general<br>
<span>… if c14n was defined on RDF 1.2 with triple terms, unstar would not be necessary.</span></p>
<p><cite>pfps:</cite> as far as I'm concerned, this group does not need to do anything about c14n.<br>
<span>… We could be nice, but not spend to much effort about this.</span><br>
<span>… We could say "RDF-CANON should be extended to RDF 1.2. In the mean time, use an unstar mapping that essentially uses reification."</span></p>
<p><cite>&lt;pfps&gt;</cite> This should be specified a bit more precisely.</p>
<p><cite>tl:</cite> I agree that we should provide a reification-based unstar mapping, that does not require named graphs.<br>
<span>… Several variants are described in the github issue, I think the reification-based is the less controversial.</span></p>
<p><cite>niklasl:</cite> we need to know what the type of triple terms is.</p>
<p><cite>gkellogg_:</cite> although the unstar is necessary for c14n, there are probably other cases that require it.<br>
<span>… I think round-tripping is important.</span><br>
<span>… We must consider SPARQL as well, being able to run it on an RDF 1.1 database.</span></p>
<p><cite>&lt;niklasl&gt;</cite> +1 to "unstar a query" too (to put annotation in existing graph stores and use it in production)</p>
<p><cite>gkellogg_:</cite> It would also be interesting to be able to turn standard reification into triple terms.</p>
<p><cite>&lt;niklasl&gt;</cite> (noting that I have no problem "unstarring queries" manually)</p>
<p><cite>ora:</cite> could be an issue if there is a mix of old-style and new-style.</p>
<p>gkellogg_; we need to decide if that's a problem.</p>
<p><cite>&lt;niklasl&gt;</cite> +1 to that to (is the target 1.2 basic or 1.1)</p>
<p><cite>&lt;pfps&gt;</cite> In my opinion an unstar mapping is only to be used as an interim solution.  It does not have to support round-tripping.  It does not need to preserve non-isomorphism for RDF graphs that are not "good" RDF 1.2 graphs (misusing rdf:reifies, misusing triple terms, not well-formed if we still have this).</p>
<p><cite>AndyS:</cite> we need to decide if we are turning RDF 1.2 Full into RDF 1.2 Classic or into RDF 1.1. RDF 1.2 has text direction, RDF 1.1 does not. The algorithm would not be the same.</p>
<p><cite>gkellogg_:</cite> I claim that base direction would not impact c14n.<br>
<span>… In the sense that it can easily be integrated in the existing algorithm.</span></p>
<p><cite>pfps:</cite> given that there is more than c14n does not handle, I say we don't do anything.</p>
<p><cite>ora:</cite> who's gonna get hurt if we don't do anything?</p>
<p><cite>gkellogg_:</cite> the c14n algorithm is not so complicated, but the proof that it's correct are.<br>
<span>… It is not likely that a new group is going to be chartered to do a new algorithm.</span><br>
<span>… If we don't provide an unstar mapping, the RCH group would probably provide one.</span><br>
<span>… We do have the Full and Basic profiles.</span></p>
<p><cite>pchampin:</cite> I simpathise with pfps position, but agree that if we don't do it someone else will.<br>
<span>… I believe there are other use cases for un-star.</span><br>
<span>… This provides a useful way to go between profiles.</span></p>
<p><cite>&lt;tl&gt;</cite> I think we should do it (backwards compatibility, and illustrate the meaning of triple terms), and re @AndyS its target should be RDF 1.1</p>
<p><cite>AndyS:</cite> I think we should do something, especially since the CG report already mentions an unstar mapping.</p>
<p><cite>&lt;niklasl&gt;</cite> This issue has mutated a bit, and should perhaps be retitled again, to e.g. "Define an unstarred form into RDF 1.1(?)" (to cater for everything specified and/or deployed, including C14N and reasoners, that will not upgrade to RDF 1.2 overnight)?</p>
<p><cite>AndyS:</cite> does not have to be a big thing, can be non-normative.</p>
<p><cite>ora:</cite> if there are other use-cases for this than c14n, we run the risk that other people do it differently.<br>
<span>… We are in the best position to do it.</span><br>
<span>… What are our use-cases? Is it someone who wants to use RDF 1.2 with their RDF 1.1 implementation?</span></p>
<p><cite>&lt;Zakim&gt;</cite> pfps, you wanted to point out that <a href="https://www.w3.org/TR/rdf-canon/" rel="noopener noreferrer">https://<wbr>www.w3.org/<wbr>TR/<wbr>rdf-canon/</a> has flas</p>
<p><cite>ora:</cite> Other question is round-tripping.</p>
<p><cite>pfps:</cite> the RDF-CANON has a big problem, it is unclear about what its input is.<br>
<span>… The input is an n-quads documents, but in other places it talks about the input dataset.</span><br>
<span>… I don't want us to get involved in this.</span><br>
<span>… It is not up to us to fix that problem.</span></p>
<p><cite>gkellogg_:</cite> you seem to be asserting that it is a flawed recommendation.</p>
<p><cite>&lt;pfps&gt;</cite> From RDF canonicalization "This document outlines an algorithm for generating a canonical serialization of an RDF dataset given an RDF dataset as input."</p>
<p><cite>gkellogg_:</cite> It uses n-quads as an input to leverage the blank node labels.</p>
<p><cite>&lt;pfps&gt;</cite> From the docuemnt "input dataset</p>
<p><cite>&lt;pfps&gt;</cite>     The abstract RDF dataset that is provided as input to the algorithm."</p>
<p><cite>gkellogg_:</cite> It was reviewed and I think it is an adequate recommdnation.</p>
<p><cite>&lt;pchampin&gt;</cite> s/recommndnation/recommendation</p>
<p><cite>&lt;pfps&gt;</cite> The algorithm, however, starts with an N-quads document.</p>
<p><cite>niklasl:</cite> I think the argument about migrating from RDF 1.1 is a good one.</p>
<p><cite>&lt;tl&gt;</cite> +1 to niklasl</p>
<p><cite>pchampin:</cite> I think we should use RDF 1.2 Classic as the baseline, not RDF 1.1.</p>
<p><cite>niklasl:</cite> I sympathise with this position.</p>
<p><cite>&lt;william_vw&gt;</cite> I assume RDF 1.2 classic means without triple terms?</p>
<p><cite>niklasl:</cite> There are ways in JSON-LD 1.1 to encode base direction in RDF 1.1, but there are two options, none of them ideal.</p>
<p><cite>ora:</cite> what would it require to deal with base direction in RDF 1.1?</p>
<p><cite>&lt;tl&gt;</cite> @william_vw yes</p>
<p><cite>&lt;william_vw&gt;</cite> @tl thanks</p>
<p><cite>gkellogg_:</cite> we would probably go with one of the JSON-LD options, probably the i18n datatype.</p>
<p><cite>&lt;niklasl&gt;</cite> I'd agree on i18n datatype version of lang-literals-with-dir.</p>
<p><cite>ora:</cite> I don't like the idea to deem this somebody else's problem<br>
<span>… Depends how much work we want to put into it, pieces of solution already exist.</span></p>
<p><cite>pchampin:</cite> I agree that we need to do something. But, let's be clear on what we're trying to solve.<br>
<span>… I'm worried about using RDF 1.1 as a baseline.</span><br>
<span>… It's also useful for the "inter-profile" operation.</span><br>
<span>… If we agree that this is the scope, I can take it into account in my PR.</span></p>
<p><cite>ora:</cite> it would solve the c14n "for free". Ultimately RDF-CANON has to be adapted to RDF 1.2, but we know it may take time.</p>
<p><cite>&lt;ora&gt;</cite> STRAWPOLL: Should we include an un-star algorithm in the 1.2 spec? (some questions remain...)</p>
<p><cite>&lt;pfps&gt;</cite> -1</p>
<p><cite>&lt;niklasl&gt;</cite> +1</p>
<p><cite>&lt;gkellogg_&gt;</cite> +1</p>
<p><cite>&lt;tl&gt;</cite> +1</p>
<p><cite>&lt;ora&gt;</cite> +1</p>
<p><cite>&lt;william_vw&gt;</cite> +1</p>
<p><cite>&lt;MacTed&gt;</cite> +1</p>
<p><cite>&lt;olaf&gt;</cite> +1</p>
<p><cite>&lt;pchampin&gt;</cite> +1</p>
<p><cite>&lt;AndyS&gt;</cite> +1</p>
<p><cite>&lt;Dominik_T&gt;</cite> -0.5</p>
<p><cite>gtw:</cite> what would normative mean? not all implementations need the unstar mapping?</p>
<p><cite>pchampin:</cite> it needs to be normative if other specs want to normatively refer to it.</p>
<p><cite>gkellogg_:</cite> implementations can be compliant without passing all the tests</p>
<p><cite>&lt;gtw&gt;</cite> +1</p>
<p><cite>gtw:</cite> the manifest vocabulary allows us to flag tests as requiring certain features</p>
<p><cite>tl:</cite> it seems like a straightforward implementation, not putting unreasonable burden on implementations, so let's make it normative</p>
<p><cite>&lt;niklasl&gt;</cite> I like this direction (not requiring 1.2 implementations to have an unstar operation plus adding a new set of tests for this feature (usable to opt-in for those who support it)).</p>
<p><cite>ora:</cite> if we didn't make it normative, that would nudge the RCH group to do the work properly</p>
<p><cite>AndyS:</cite> the more tests we put, the more work that is, and the less likely it is that implementations migrate</p>
<p><cite>ora:</cite> I don't know the answer to this normative/non-normative thing. If we make a non-normative section, then people will hopefully not make their own.<br>
<span>… If some days the RCH group makes a 1.2, this is an interim solution.</span></p>
<p><cite>pchampin:</cite> in the interest of time, I suggest we defer the nonrmative/non-normative question until we have a PR to discuss on</p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/rdf-star-wg/issues/114" rel="noopener noreferrer">Issue 114</a></strong> Un-star operation to support RDF Dataset Canonicalization? (by niklasl) [needs discussion] [discuss-f2f]</p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/sparql-service-description/issues/26" rel="noopener noreferrer">Issue 26</a></strong> shape languages and validation features (by phtyson) [discuss-f2f] [needs discussion]</p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/rdf-star-wg/issues/127" rel="noopener noreferrer">Issue 127</a></strong> Material about `rdf:ReificationProperty` (by afs) [needs discussion] [discuss-f2f]</p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/sparql-query/issues/156" rel="noopener noreferrer">Issue 156</a></strong> Addressing SPARQL EXISTS errata (by afs) [discuss-f2f] [needs discussion]</p>
<p><cite>&lt;niklasl&gt;</cite> I see no big risk in just making it informative. (Famous last words; but... In practice, for "reasons", lots of implementations aren't 1:1 with what's normative.)</p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/rdf-semantics/issues/49" rel="noopener noreferrer">Issue 49</a></strong> Define an interpretation of Triple Terms (by niklasl) [discuss-f2f] [needs discussion]</p>
<p><cite>&lt;niklasl&gt;</cite> I see no *big* risk in just making it informative. (Famous last words; but... In practice, for "reasons", lots of implementations aren't 1:1 with what's normative.)</p>

<hr /></details>

**Comment by pchampin** [comment]
This was discussed during the [rdf-star meeting on 26 September 2024](https://www.w3.org/2024/09/26-rdf-star-minutes.html#t05).

<details><summary><i>View the transcript</i></summary>

<h3>Addressing SPARQL EXISTS errata <a href="https://github.com/w3c/sparql-query/issues/156" rel="noopener noreferrer">4</a></h3>
<p><cite>ora:</cite> Are there people fine with the current syntax?</p>
<p><cite>ora:</cite> In any case, chairs will discuss this, let's move on</p>
<p><cite>AndyS:</cite> [about SPARQL EXISTS] There are two proposals</p>
<p><cite>AndyS:</cite> 1. substitution based on various existing errata</p>
<p><cite>AndyS:</cite> 2. an other one based on ANTIJOIN. We already have MINUS. Except the behavior with disjoin domain. But outside of it it's ANTIJOIN</p>
<p><cite>AndyS:</cite> On an other note, there are other things that might go to SPARQL like LATERAL that can be based on substitution. And pure form of anti join and semi join</p>
<p><cite>AndyS:</cite> It's a possibility to move these additions (LATERAL, anti join...) to sparql dev</p>
<p><cite>pchampin:</cite> we would add more subtly differences between operators like FILTER NOT EXISTS vs MINUS</p>
<p><cite>pchampin:</cite> Your point of having multiple ways might create problems</p>
<p><cite>ora:</cite> SPARQL spec spends a bit of time presenting this difference</p>
<p><cite>AndyS:</cite> It was quite contentious in SPARQL 1.1</p>
<p><cite>&lt;pchampin&gt;</cite> I'm more than happy to let the editors decide on that</p>
<p><cite>AndyS:</cite> I am not aware of any outgoing opinion, I think it ends up to a choice on which way to go</p>
<p><cite>tl:</cite> is it related to triple terms in any way of is it a SPARQL errata</p>
<p><cite>AndyS:</cite> it has nothing to do with triple terms</p>
<p><cite>tl:</cite> what is the criteria of SPARQL errata to discuss now?</p>
<p><cite>tl:</cite> it's a central issue, is that the argument?</p>
<p><cite>pfps:</cite> There are a bunch of problems with SPARQL, the ones with EXIST are the biggies</p>
<p><cite>pfps:</cite> They end up splitting the SPARQL implementation space</p>
<p><cite>pfps:</cite> The decision that has to be made is to move SPARQL EXIST toward a more database-like implementation and keep it more consistent with the existing</p>
<p><cite>AndyS:</cite> The current implementation is present in SQL with correlated subqueries</p>
<p><cite>pfps:</cite> if you use the semi/anti join interepretation of EXISTS you change SPARQL more than the other option</p>
<p><cite>pfps:</cite> In the end people who will see and understand the differences are very few</p>
<p><cite>ora:</cite> I would like to know preferences</p>
<p><cite>AndyS:</cite> My preference is for substitution and applying errata (option 1)</p>
<p><cite>pfps:</cite> I don't have much of a horse in this race</p>
<p><cite>pfps:</cite> Idealy I would love to get more SPARQL developers on board</p>
<p><cite>ora:</cite> we could talk outside of the group</p>
<p><cite>ktk:</cite> I reached out to stardog but not got an input</p>
<p><cite>gtw:</cite> I am not sure much value to reach out to more developers. sparql-dev has been opened for a long time</p>
<p><cite>&lt;pchampin&gt;</cite> Tpt: I have a signicant preference for option 1; option 2 is basically equivalent to MINUS</p>
<p><cite>pfps:</cite> One way to check the issue would be to pull some tests</p>
<p><cite>&lt;pfps&gt;</cite> which PR?</p>
<p><cite>&lt;gkellogg&gt;</cite> <a href="https://github.com/w3c/rdf-tests/issues/42" rel="noopener noreferrer">w3c/<wbr>rdf-tests#42</a></p>
<p><cite>&lt;gb&gt;</cite> <strong><a href="https://github.com/w3c/rdf-tests/issues/42" rel="noopener noreferrer">Issue 42</a></strong> tests to document current definition of EXISTS in SPARQL (by pfps) [SPARQL]</p>
<p><cite>&lt;gkellogg&gt;</cite> <a href="https://github.com/w3c/rdf-tests/pull/43" rel="noopener noreferrer">w3c/<wbr>rdf-tests#43</a></p>
<p><cite>&lt;gb&gt;</cite> <a href="https://github.com/w3c/rdf-tests/pull/43" rel="noopener noreferrer">CLOSED Pull Request 43 Add tests to document current definition of EXISTS (by pfps)</a></p>
<p><cite>ora:</cite> Whatever solutions we pick, someone will ask why we pick it</p>
<p><cite>AndyS:</cite> picking sustitution breaks the least queries</p>
<p><cite>ora:</cite> That seems to me a as good reason as any, let's make a decision</p>
<p><cite>tl:</cite> I would like to ask james about it</p>
<p><cite>ora:</cite> Let's vote on it next Thursday</p>
<p><cite>ora:</cite> Let's do it</p>

<hr /></details>

**Comment by VladimirAlexiev** [comment]
@domel sent us this "questionnaire" by email. I'll paste it below, because I find it a useful roadmap to this topic.

The RDF-star Working Group is currently addressing issues related to updating the semantics of SPARQL EXISTS, specifically regarding:
- Certain uses of EXISTS being undefined during evaluation,
- Substitution occurring where definitions apply only to variables,
- Blank nodes being substituted into BGPs and acting as variables,
- Substitution potentially flipping MINUS into its disjoint-domain case,
- Substitution impacting disconnected variables.

Would you be able to provide your insights on these matters

**Comment by afs** [comment]
> @domel sent us this "questionnaire" by email. I'll paste it below, because I find it a useful roadmap to this topic.
> 
> The RDF-star Working Group is currently addressing issues related to updating the semantics of SPARQL EXISTS, specifically regarding:
> 
> * Certain uses of EXISTS being undefined during evaluation,
> * Substitution occurring where definitions apply only to variables,
> * Blank nodes being substituted into BGPs and acting as variables,
> * Substitution potentially flipping MINUS into its disjoint-domain case,
> * Substitution impacting disconnected variables.
> 
> Would you be able to provide your insights on these matters

This section of SEP-0007 expands on these points:
https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md#identified-issues

**Comment by hannahbast** [comment]
Dear all, Hannah (@hannahbast) and Johannes (@joka921) here from the University of Freiburg and developers of [QLever](https://github.com/ad-freiburg/qlever). Fascinating and important question. There is a lot to untangle here, so first

**TLDR: In border cases regarding the semantics of a complex query, we recommend to follow the inner logic of the query language and not what a user thinks the query should do. Following that, we recommend to define `EXISTS` via a standard semijoin. That being said, the SPARQL standard currently misses some important features, but that is a different topic and should not be conflated with the topic under discussion here.**

Here is the long version:

1. There are two principle ways to determine what the semantics of a certain query should be. One is based on the internal logic of the query language. The other is based on what users think a certain query does or should do. The two should typically coincide, otherwise something would be very wrong with the design of the query language. But the definition of *anything* non-trivial is bound to have some unintuitive border bases.

2. The definition of the semantics of `EXISTS` has several such border cases, notably when the right side uses filters on variables that only occur on the left side. It is aggravated by the fact that most users don't really think about `EXISTS` as the [functional form](https://www.w3.org/TR/sparql11-query/#func-forms) it is, but about its usage in combination with a filter, as in `FILTER NOT EXISTS`, and then wondering why on earth SPARQL has both `MINUS` and `FILTER NOT EXISTS`, which are very similar but not equal. But that is really just a side-effect of `EXISTS`. So better think about `EXISTS` in isolation, like in this simple example query: [three people with their name and whether they have a birth date in Wikidata or not](https://w.wiki/BQ5Y).

6. Most SPARQL features translate to some form of join operation when processing the query. For example, two graph patterns with joint variables can be processed by a standard **join** operation: the result is all pairs of solutions from the left and from the right, where the joint variables have the same value. One graph pattern `MINUS` another can be processed by an **antijoin**: the result is only those solutions from the left, which don't match one from the right. Similarly, `OPTIONAL` translates to a **left outer join**.

7. All these join operations have an important property: the left and right side can be evaluated **independently**, and the computational complexity is **linear** in the size of the input plus the size of the output (modulo some details related to `UNDEF` values omitted here).

8. Following that logic, it would make a lot of sense to define `EXISTS` via a **semijoin**: That is, each solution on the left evaluates to true if and only if it matches a solution on the right. This includes the special case, where there are no variables in common: then all solutions on the right match, just like the would in a normal join, and each solution on the left evaluates to true (and this also explains the difference between `FILTER NOT EXISTS` and `MINUS`).

9. Unfortunately, the standard didn't define `EXISTS` that way, but chose to define it via **substitution**. Substitution is conceptually different from a join-like operation because it is a **quadratic** algorithm: for each solution on the left, perform the corresponding substitution on the right and evaluate it. It is true that one can reduce this to ordinary joins in many cases (a practice called de-correlation or unnesting, [here is a good paper about it](https://btw-2015.informatik.uni-hamburg.de/res/proceedings/Hauptband/Wiss/Neumann-Unnesting_Arbitrary_Querie.pdf)). But the current definition does go against the inner logic of all the other SPARQL constructs, as sketched above. And you can already tell that something is fishy by looking at the standard, where "substitute" is defined solely for the purpose of defining `EXISTS` and used nowhere else: https://www.w3.org/TR/sparql11-query/#sparqlAlgebraEval . The many [issues with `EXISTS`](https://github.com/w3c/sparql-query/issues?q=is%3Aissue+is%3Aopen+EXISTS) are a direct consequence of this special treatment.

12. All that being said, there are important features missing from SPARQL. One notable example is queries like [the highest peak of each country](https://qlever.cs.uni-freiburg.de/wikidata/FMZ8Sr?exec=true), where you have a `GROUP BY` (for each country, all the peaks), an aggregation (find the maximum height per country), but then want the value from another column corresponding to the result of the aggregation (the peak which has the maximum height). SPARQL currently forces you to write a very unnatural query for this, where parts of the query have to be repeated in a subquery. Another example are parametrized queries like [all streets in region X](https://qlever.cs.uni-freiburg.de/osm-planet/53wuct?exec=true) (click on "Map view" to see the result on a map), where you want to define one or more parameters at the top, which are then used in the rest of the query. This works for simple queries, but with current SPARQL breaks as soon as you have subqueries or nested group graph patterns using the query parameters, which is confusing for many users.

13. These missing features are related to the "substitute" semantics and to nested queries and should definitely be addressed and we do have recommendations for that. But **we strongly advise against conflating this with the current discussion**, which is about `EXISTS` and not about all kinds of missing features. When implementing new features that depart from the join logic inherent to the current version of the SPARQL standard, we would advise to use new keywords (`LATERAL` being an example) in order to make the departure from the standard logic explicit.

**Comment by pchampin** [comment]
This was discussed in [today's meeting](https://www.w3.org/2024/10/03-rdf-star-minutes.html#5d28)

**Comment by ktk** [comment]
We need to discuss a [Task Force](https://www.w3.org/2025/03/27-rdf-star-minutes.html#ee16) for this

**Comment by w3cbot** [comment]
This was discussed during the [#rdf-star meeting on 10 April 2025](https://www.w3.org/2025/04/10-rdf-star-minutes.html#63db).

<details><summary><i>View the transcript</i></summary>

<h3>Task Force for SPARQL EXISTS <a href="https://github.com/w3c/sparql-query/issues/156" rel="noopener noreferrer">3</a></h3>
<p><cite>ora:</cite> Idea is to prepare for EXISTS - not a priority for the whole WG at this time.<br>
<span>… does the group agrees?</span><br>
<span>… who will chair?</span></p>
<p><cite>&lt;AndyS&gt;</cite> s/agress/agree/</p>
<p><cite>ora:</cite> james - would you chair this?</p>
<p><cite>james:</cite> insufficient experience of W3C processes</p>
<p><cite>ora:</cite> it involves scheduling and making the TF moves forward</p>
<p><cite>gkellogg:</cite> Agree to TF and maybe other items in the WG for subsets of the participants<br>
<span>… creates a place for discussions. I would participate.</span></p>
<p><cite>james:</cite> sub-group decide chair?</p>
<p><cite>Andys:</cite> I can schedule a first meeting</p>
<p><cite>&lt;TallTed&gt;</cite> TF(s) will let focused conversation(s) take place in parallel with main group without consuming main group time. I don't think I will have the time to do much if any more than participate (which I will *try* to do).</p>
<p><cite>ktk:</cite> can external people participate?</p>
<p><cite>&lt;TallTed&gt;</cite> TF participants must be WG members, whether as W3CMember reps or as IEs</p>
<p><cite>james:</cite> certainly agree for external participants</p>
<p><cite>pchampin:</cite> chairs can invite (and to WG meetings)<br>
<span>… need to check if a member org can specific a person outside of their org.</span></p>
<p><cite>&lt;TallTed&gt;</cite> IP issues can quickly become quite hairy.</p>
<p><cite>tallted:</cite> depends on their input (IP issues need care)<br>
<span>… if they are contributing significantly, should do the IP side.</span></p>
<p><cite>ora:</cite> adjourned</p>

<hr /></details>

**Comment by afs** [comment]
> schedule a first meeting

WG members - please let me know (by direct email) of your interest and availability.

**Comment by pfps** [comment]
There appears to be two different ways of fixing EXISTS and they produce different answers.

One method adds the current bindings to the FILTER expression, roughly as if a VALUES expression was preprended to it, but evaluates the result normally.  This means the result of any subqueries is the same for all the bindings so they need only be evaluated once.

The other method is much more involved, and is described in https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md  In this method the current bindings are pushed into subqueries with the result that they may need to be evaluated multiple times.  (It may be possible to optimize the evalation so that the subqueries do not need to be completely re-evaluated for each binding.)

A query that shows a difference between the two ways is:

```
PREFIX ex: <ex:>
SELECT ?v WHERE {
BIND ( ex:a AS ?v )
FILTER EXISTS { SELECT ?v WHERE { FILTER (?v = ex:a) } }
}
```

Under the first method this query produces no answers because no binding is available for ?v in the embedded query.  Under the second method an answer is produced because the current bindings are pushed all the way down in the expression.   (This is assuming that there is a BGP in the subquery, otherwise there is a bug in the second method.)

Currently MilleniumDB and Blazegraph produce no answer sets for this query.
QLever and www.sparql.org do produce an answer set.

**Comment by TallTed** [comment]
For what it's worth, current Virtuoso also produces a single row result set. ( See [query](https://demo.openlinksw.com/sparql?default-graph-uri=&qtxt=PREFIX+ex%3A+%3Cex%3A%3E%0D%0ASELECT+%3Fv+WHERE+%7B%0D%0ABIND+%28+ex%3Aa+AS+%3Fv+%29%0D%0AFILTER+EXISTS+%7B+SELECT+%3Fv+WHERE+%7B+FILTER+%28%3Fv+%3D+ex%3Aa%29+%7D+%7D%0D%0A%7D%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0), and [result](https://demo.openlinksw.com/sparql?default-graph-uri=&query=PREFIX+ex%3A+%3Cex%3A%3E%0D%0ASELECT+%3Fv+WHERE+%7B%0D%0ABIND+%28+ex%3Aa+AS+%3Fv+%29%0D%0AFILTER+EXISTS+%7B+SELECT+%3Fv+WHERE+%7B+FILTER+%28%3Fv+%3D+ex%3Aa%29+%7D+%7D%0D%0A%7D%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0).)

**Comment by afs** [comment]
@pfps -- What does the SPARQL 1.1 say?

"the pattern formed by replacing every occurrence of a variable v in pattern by μ(v) for each v in dom(μ)"

Aside from the projection of ?v (separate issue that you have [already identified](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md#issue-2-substitution-happens-where-definitions-are-only-for-variables), 

What about:
```
PREFIX ex: <ex:>
SELECT ?v WHERE {
BIND ( ex:a AS ?v )
FILTER EXISTS { SELECT * WHERE { FILTER (?v = ex:a) } }
}
```
```
PREFIX ex: <ex:>
SELECT ?v WHERE {
BIND ( ex:a AS ?v )
FILTER EXISTS { FILTER (?v = ex:a) }
}
```

**Comment by afs** [comment]
[PR 177](https://pr-preview.s3.amazonaws.com/w3c/sparql-query/pull/177.html#values-insertion-and-exists) has one way in which `substitute` might be fixed.

The principle is that during execution of the row being filtered, variables have their value constrained to the binding in the row.

The PR describes this by a process of rewriting the algebra on a per-row basis. It is not a required implementation approach.

It is not the only way to communicate the effect. An alternative possibility is to a new algebra operator which evaluates to the row being filtered. This would make for a rewrite a static (algebra build time) step.

**Comment by afs** [comment]
I've setup a repo for tests and notes: https://github.com/afs/SPARQL-exists

In it, there is the [example query](https://github.com/w3c/sparql-query/issues/156#issuecomment-2824827421), and [variations](https://github.com/w3c/sparql-query/issues/156#issuecomment-2842525316), together with `NOT EXISTS` versions and `NOT EXISTS` replaced by `MINUS`.

https://github.com/afs/SPARQL-exists/tree/main/tests/exists-filter

**Comment by lisp** [comment]
there is now a [pull request](https://github.com/w3c/sparql-query/pull/211) to explore minimal changes to the recommendation to support the interpretation, that it should be implemented analogous to nested loop bgp mechanisms.

the change clarifies that patterns are _not_ to be instantiated, which eliminates the errors otherwise associated the blank nodes.  
we have followed this interpretation from our first implementation over a decade ago.  
the implementation produces results which differed with those proposed during the sparql-dev working group discussions only due to differences in variable scoping rules.

**Comment by pfps** [comment]
It would be useful to describe how this proposal would change common uses of EXISTS.  For example what happens with

```
ex:a ex:b ex:c .
ex:k ex:d ex:f .
```

```
SELECT ?x WHERE {
  ex:a ex:b ?x .
  FILTER NOT EXISTS { ?x ex:d ex:f . }
}
```

**Comment by afs** [comment]
> One method adds the current bindings to the FILTER expression, roughly as if a VALUES expression was prepended to it, but evaluates the resulting expression normally.

Prepend inside or outside the `{...}`?

The second method can be seen as prepending inside the `{...}`.

In the syntax to algebra algorithm of SPARQL, there is also always an empty BGP at the start of a group graph pattern. The ["simplification" step](https://www.w3.org/TR/sparql11-query/#sparqlSimplification) in the spec replaces "join(Z, X)" and "join(X, Z)" with X for clarity.

The values insertion process includes a VALUES block of a single row (actually `Table(μ)`), at the start of all group graph patterns.

There is a binding for `?v`. In
```
{ FILTER (?v = ex:a) }
```
there is no join, and `Z` is still there.

So it becomes (algebra)
```
(filter (?v = ex:a)
    (join Z Table(μ))
)
```
Filter (algebra) always filers something.

> Under the first method this query produces no answers because no binding is available for ?v in the embedded query.

Maybe because it is outside the `{...}`?

`{ VALUES(μ) FILTER (?v = ex:a) }` would work, but `VALUES(μ) { FILTER (?v = ex:a) }` is not prepending immediately before the `FILTER`,

**Comment by afs** [comment]
Added blank node tests ([issue 3](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md#issue-3-blank-nodes-substituted-into-bgps-act-as-variables)).

**Comment by hannahbast** [comment]
Dear all, I just reread my [comment from 03.10.2024](https://github.com/w3c/sparql-query/issues/156#issuecomment-2390802086) and still feel the same way. In case this got forgotten, here are the main points again; see the comment for examples and all the details:

1. When in doubt about the semantics of a query, follow the inner logic of the query language and not what users think the result should be
2. According to the inner logic of the query language, the natural definition of `EXISTS` is via a standard semijoin and **not** via substitution
3. I understand the desire and need for substitution semantics in SPARQL, but there should be separate keywords for this

**Comment by pfps** [comment]
> [@pfps](https://github.com/pfps) -- What does the SPARQL 1.1 say?
> 
> "the pattern formed by replacing every occurrence of a variable v in pattern by μ(v) for each v in dom(μ)"
> 
> Aside from the projection of ?v (separate issue that you have [already identified](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0007/sep-0007.md#issue-2-substitution-happens-where-definitions-are-only-for-variables),
> 
> What about:
> 
> ```
> PREFIX ex: <ex:>
> SELECT ?v WHERE {
> BIND ( ex:a AS ?v )
> FILTER EXISTS { SELECT * WHERE { FILTER (?v = ex:a) } }
> }
> ```

My view is that embedded queries should be treated inside EXISTS just as they are outside exists.

> ```
> PREFIX ex: <ex:>
> SELECT ?v WHERE {
> BIND ( ex:a AS ?v )
> FILTER EXISTS { FILTER (?v = ex:a) }
> }
> ```

But that in-scope bindings are available at the top level of the EXISTS pattern.

So these two have different results, just like

```
PREFIX ex: <ex:>
SELECT ?v WHERE {
 BIND (ex:a AS ?v )
 FILTER (?v = ex:a)
}
```

and

```
PREFIX ex: <ex:>
SELECT ?v WHERE {
 BIND (ex:a AS ?v )
 { SELECT * WHERE { FILTER (?v = ex:a) } } 
}
```

do.

**Comment by kasei** [comment]
@hannahbast –

> 2\. According to the inner logic of the query language, the natural definition of EXISTS is via a standard semijoin and not via substitution
> 3\. I understand the desire and need for substitution semantics in SPARQL, but there should be separate keywords for this

My recollection of the 1.1 WG is that (NOT-)EXISTS was mostly motivated by the desire for negation support. In that context, I think `(NOT) EXISTS` *is* the "separate keyword" for substitution semantics, while `MINUS` provided the join-based negation semantics. (And we didn't get a semi-join counterpart to the negative `MINUS`).

**Comment by afs** [comment]
> Dear all, I just reread my [comment from 03.10.2024](https://github.com/w3c/sparql-query/issues/156#issuecomment-2390802086) and still feel the same way. In case this got forgotten,

@hannahbast - The comment has not been forgotten.

Johannes presented a summary in the meeting on Friday (May 2nd)

https://www.w3.org/2025/05/02-rdf-star-minutes.html#10000

**Comment by afs** [comment]
The topic for the SPARQL TF meetign this week is agree on the 5 issues, with the understanding that new issues may be found. The expectation for a new issue is that it needs clearly stating what the problem is, with evidence, such as query, data and results.

WG members - the meetings should now be appearing on your W3C calendar and also in the working group upcoming meetings.

**Comment by lisp** [comment]
attached here is a script to run the tests from [https://github.com/afs/SPARQL-exists/tree/main/tests](https://github.com/afs/SPARQL-exists/tree/main/tests) against a [repository](https://dydra.com/sparql-12/exists-tests) in dydra.

the data is present in distinct graphs in a single dataset.

[run_tests.txt](https://github.com/user-attachments/files/20236329/run_tests.txt)

**Comment by afs** [comment]
I've reorganised the [tests area](https://github.com/afs/SPARQL-exists) with a set of tests per issue as suggested at the last meeting. (There is overlap because the issues are not independent.)

Also - I've added a description of using an algebra function to PR #177. The function evaluates to the current solution mapping (mentioned in [comment](https://github.com/w3c/sparql-query/issues/156#issuecomment-2842843762).

The spec description now does imply changing the algebra for query execution once algebra expression is initially constructed.

**Comment by lisp** [comment]
what do test harnesses do with results from the broader sparql test suit.
i leaned toward adjusting the result comparison to rewrite plank node labels, but have bad conscience about that approach.

**Comment by afs** [comment]
> At least one of the tests returns a bnode.

Which? Is is one of the exists-N-* tests or the bnode specific area?

> Some engines don't return the bnode ID unchanged

All blank node labels are scoped to the document.

Do a check for bnode isomorphism.

> what do test harnesses do with results from the broader sparql test suit.

Such a check is required for the wider SPARQL test suite.

**Comment by pfps** [comment]
> > At least one of the tests returns a bnode.
> 
> Which? Is it one of the exists-N-* tests or the bnode specific area?

#exists-bgp-2 in exists-3-bgp

> > Some engines don't return the bnode ID unchanged
> 
> All blank node labels are scoped to the document.
> 
> Do a check for bnode isomorphism.

I don't think that that works.   If a result has only a blank node ID, how can you tell which one it is in the graph?  The label is scoped to the result set, not the graph.

**Comment by afs** [comment]
> If a result has only a blank node, how can you tell which one it is in the graph?

The two result sets must be the same by isomorphism.
It may take designing the data and test carefully but the tests "exists-N-*" are yes/no filters with either rows or no rows.

**Comment by pfps** [comment]
> > If a result has only a blank node, how can you tell which one it is in the graph?
> 
> The two result sets must be the same by isomorphism. It may take designing the data and test carefully but the tests "exists-N-*" are yes/no filters with either rows or no rows.

I don't think that adequate care was taken.

Consider the following:

Data:

PREFIX ex: <http://example.com/>
_:a ex:b ex:c .
_:b ex:b ex:d .

Query:

PREFIX ex: <http://example.com/>
SELECT ?x WHERE {
 ?x ex:b ex:c .

Specified Result:

?x
_:a

Actual Result:

?x
_:b

How can one determine whether a result is correct without examining both the data and the query?

If the data has only one blank node then maybe it is easier, but that would require examining the data, not just the specified and actual results.

With something like:

Data:

PREFIX ex: <http://example.com/>
_:a ex:b ex:c .
_:a ex:id "a" .
_:b ex:b ex:d .
_:b ex:id "b" .

Query:

PREFIX ex: <http://example.com/>
SELECT ?id WHERE {
 ?x ex:b ex:c .
 ?x ex:id ?id

Specified Result:

?id
a


The data and query have been arranged so that there is one correct answer.  Code that determines whether the answer is correct only needs to look at the answer.

**Comment by afs** [comment]
> the beginning, not in subqueries.

Page 3:

```
EXIST { ?person foaf:name ?name . }
EXIST { pattern }
```
becomes
```
{ VALUES ?person { :Bill } ?person foaf:name ?name . }
{ VALUES ?person { :Bill } pattern }
```
which is a join of the VALUES with the result of the pattern. We only need row/no-row so that is equivalent to semijoin/antijoin.

The pattern may have structure more than a BGP, subqueries are one case, but there are other structures which give depth -- e.g. OPTIONAL, UNION, even `{}` and FILTER -- which cause deviation from the exist spec on existing well-defined cases.

**Comment by pfps** [comment]
I thought that the semijoin/antijoin would give no answers for 

Data:

ex:a ex:b ex:c .

Query:

SELECT ?x WHERE {
  ?x ex:b ex:c .
  FILTER EXISTS { FILTER ( ?x = ex:a ) }
}

**Comment by afs** [comment]
Depends on how much work is done in the mapping decision.

Matching the result sets
```
  ?x   ?y
 _:A  _:B
 _:A  _:C 
 _:C  _:D
```

```
  ?x   ?y
 _:X  _:Z
 _:X  _:Y
 _:Z  _:W
```
has a choice at matching (A,B) to (X,Y) or (X,Z), broken by (C, D)/(Z,W).

**Comment by TallTed** [comment]
@pfps — Can you please add links to `PLEASE IGNORE- SEE BELOW:` and add some markup to the rest of that comment that help make plain what is to be ignored, and what is to be seen below? I would also strongly advise adding code fences to the SPARQL, Turtle, and similar sections (for instance, the —

PREFIX ex: <http://example.com/> 

— we see in https://github.com/w3c/sparql-query/issues/156#issuecomment-2906970355 is invalid, while I'm sure what you actually meant was the valid —
```
PREFIX ex: <http://example.com/>`
```

**Comment by pfps** [comment]
I added JSON output to the EARL reports I generated and have results for Blazegraph, MilleniumDB, QLever, and Virtuoso in the wiki for this document.

**Comment by afs** [comment]
> [@pfps](https://github.com/pfps) — I'm sure I'm just not seeing it earlier, but could you add a direct link to "the wiki for this document" in [your last comment](https://github.com/w3c/sparql-query/issues/156#issuecomment-2923482439)?

Each document is a GH repo:
https://github.com/w3c/sparql-query/wiki

**Comment by TallTed** [comment]
* The Virtuoso endpoint includes the following footer (as of today) —
  > [Virtuoso](https://virtuoso.openlinksw.com/) version 08.03.3333 (22790a5421) on Linux (x86_64-generic-linux-glibc25) Single Server Edition (378 GB total memory, 26 GB memory in use)

  — which identifies a commit hash on our internal development `git` server, as this is running Virtuoso Enterprise Edition.

* The QLever endpoint includes a menu option under `Resources` that says (as of today) —
  > Git commit 68c301a on master

  — but none of the ways I've found to see a GitHub commit based on its hash seem to think that commit exists. Maybe it's an internal commit ID hash.

* The [parent page](https://www.sparql.org/) of the Jena page to which you link says (as of today) —
  > This server is running [Apache Jena Fuseki](https://jena.apache.org/documentation/fuseki2/fuseki-main.html).

  — but that link is bad. Its [parent page](https://jena.apache.org/documentation/fuseki2/) leads to [javadoc for Fuseki](https://jena.apache.org/documentation/javadoc/fuseki2/) that suggests it's running `Fuseki Core Engine 5.3.0`.

I was not quickly able to find a way to reveal the version of either Blazegraph or MillenniumDB.

As a person familiar with software, I'm sure you're aware that these are moving targets. For instance, version 1.2.3 may have a bug that versions 1.2.2 and 1.2.4 don't; version 1.2.5 may introduce support for some feature that was lacking in 1.2.4; etc. These details are vital to analyzing and patching issues such as results of the `EXISTS()` function.

**Comment by lisp** [comment]
the subject of https://github.com/w3c/sparql-query/pull/424 includes compares an exists implementation based on lateral join with the deep injection approach.  

more information on the join criteria, the implementation, and the performance is available in a longer [description](https://observablehq.com/@datagenous/exists-via-lateral-join-summary).
